Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceSocket.cpp

```diff
@@ -0,0 +1,572 @@
+        {
+            // No callback, or the callback told us to stop reading
+            return;
+        }
+
+        // if only some of the bytes were used, then shift
+        // remaining bytes down.  Otherwise, clear the buffer.
+        sock->fBytesLeft -= notifyRead.bytesProcessed;
+        if (sock->fBytesLeft != 0) {
+            if ((sock->fBytesLeft > sizeof(sock->fBuffer))
+                || ((notifyRead.bytesProcessed + sock->fBytesLeft) > sizeof(sock->fBuffer)))
+            {
+                LogMsg(kLogError, "SocketDispatchRead error: {} {} {}",
+                       sock->fBytesLeft, notifyRead.bytes, notifyRead.bytesProcessed);
+                return;
+            }
+
+            if (notifyRead.bytesProcessed) {
+                memmove(sock->fBuffer, sock->fBuffer + notifyRead.bytesProcessed,
+                        sock->fBytesLeft);
+            }
+        }
+
+        SocketStartAsyncRead(sock);
+    });
+}
+
+static void SocketGetAddresses(AsyncSocket sock, plNetAddress* localAddr,
+                               plNetAddress* remoteAddr)
+{
+    // don't have to enter critsect or validate socket before referencing it
+    // because this routine is called before the user has a chance to close it
+
+    asio::error_code err;
+    tcp::endpoint localEndpoint = sock->fSock.local_endpoint(err);
+    if (err) {
+        LogMsg(kLogError, "Failed to get local socket endpoint: {}", err.message());
+    } else if (localEndpoint.address().is_v4()) {
+        localAddr->SetHost(localEndpoint.address().to_v4().to_bytes());
+        localAddr->SetPort(localEndpoint.port());
+    } else {
+        localAddr->Clear();
+        LogMsg(kLogError, "Local socket endpoint is not IPv4");
+    }
+
+    tcp::endpoint remoteEndpoint = sock->fSock.remote_endpoint(err);
+    if (err) {
+        LogMsg(kLogError, "Failed to get remote socket endpoint: {}", err.message());
+    } else if (remoteEndpoint.address().is_v4()) {
+        remoteAddr->SetHost(remoteEndpoint.address().to_v4().to_bytes());
+        remoteAddr->SetPort(remoteEndpoint.port());
+    } else {
+        remoteAddr->Clear();
+        LogMsg(kLogError, "Remote socket endpoint is not IPv4");
+    }
+}
+
+static bool SocketInitConnect(ConnectOperation* op)
+{
+    // This steals ownership of op->fSock
+    auto sock = std::make_unique<AsyncSocketStruct>(op);
+    
+    sock->initTimeMs    = TimeGetMs();
+
+    asio::error_code err;
+    sock->fSock.non_blocking(true, err);
+    if (err)
+        LogMsg(kLogError, "Failed to set socket to non-blocking: {}", err.message());
+
+    sock->fSock.set_option(asio::socket_base::send_buffer_size(kTcpSndBufSize), err);
+    if (err)
+        LogMsg(kLogError, "Failed to set socket send buffer size: {}", err.message());
+
+    sock->fSock.set_option(asio::socket_base::receive_buffer_size(kTcpRcvBufSize), err);
+    if (err)
+        LogMsg(kLogError, "Failed to set socket recv buffer size: {}", err.message());

```

Comment:
Disabling the two set_option calls seems to resolve the download bandwidth issues on both Linux and macOS. Using the file patcher I'm seeing download speeds of up to 30 MiB/s, improved from 800 KiB/s.

There are some threads hinting that leaving these unset will allow ASIO and the underlying network stack to automatically scale buffer size up and down. I saw these values are inherited from the NT sockets implementation. I'm not sure the side effects of not setting these sizes - and I haven't retested on Windows.

Generally I can't find any discussion online yet of why POSIX sockets and NT sockets would behave so differently. So I'm unsure why this buffer size is working fine on Windows but not elsewhere.